### ISSUE TITLE
clang-format AlignArrayOfStructures crashes when seeing clauses of different number of members

### ISSUE THREAD

**Issue by lichray** [llvm/llvm-project#51767]
|  |  |
| --- | --- |
| Bugzilla Link | [52425](https://llvm.org/bz52425) |
| Version | trunk |
| OS | All |
| Attachments | [test case](https://user-images.githubusercontent.com/433009/143762695-bd222eeb-f723-45b7-a14f-a9ba87d94e99.gz) |

## Extended Description 

Crashes when formatting

```
std::vector<std::vector<uint8_t>> var_len_seq{
    { 0x80 },
    { 0xf5 },
    { 0xc3, 0x7f },
    { 0xc3, 0xc0 },
    { 0xe1, 0x7f },
    { 0xe1, 0xc0 },
    { 0xe1, 0x81, 0x7f },
};
```

If I remove the last line, I get a bad result:

```
std::vector<std::vector<uint8_t>> var_len_seq{
    {0x80},
    { 0xf5   },
    { 0xc3,   0x7f},
    { 0xc3,   0xc0},
    { 0xe1,   0x7f},
    { 0xe1,   0xc0},
};
```

My _clang-format is

---
BasedOnStyle: Microsoft
AlignArrayOfStructures: Right

...

**Comment by mkurdej** [comment]
Possible duplicate of #51277.

Crash dump:
```
Assertion failed: idx < size(), file E:\src\llvm-project\llvm\include\llvm/ADT/SmallVector.h, line 281
PLEASE submit a bug report to https://bugs.llvm.org/ and include the crash backtrace.
Stack dump:
0.      Program arguments: e:\\src\\llvm-project\\build\\Debug\\bin\\clang-format.exe -i -style=file test.cpp
0x00007FF60B78183C (0x0000003700000016 0x0000000000000000 0x00007FF60B66A984 0x0000000000000000), HandleAbort() + 0xC bytes(s), E:\src\llvm-project\llvm\lib\Support\Windows\Signals.inc, line 408
0x00007FFF8D66BC31 (0x00007FFF00000016 0x00000037683885D0 0x0000000801000006 0xFFFFFF01000000CE), raise() + 0x441 bytes(s)
0x00007FFF8D66D889 (0x00000037683885D0 0x0000000000000240 0x00007FFF8D76DA70 0x00007FF60BD2CCC8), abort() + 0x39 bytes(s)
0x00007FFF8D6734B5 (0x00007FF60BD2CCC8 0x00007FF60BD2CC40 0x0000003700000119 0x00007FFF8D65189B), _get_wide_winmain_command_line() + 0x2515 bytes(s)
0x00007FFF8D673027 (0x00007FF60BD2CCC8 0x00007FF60BD2CC40 0x0000000000000119 0x0000003768389E10), _get_wide_winmain_command_line() + 0x2087 bytes(s)
0x00007FFF8D671091 (0x00007FF60BD2CCC8 0x00007FF60BD2CC40 0x0000003700000119 0x00007FF60B973D4A), _get_wide_winmain_command_line() + 0xF1 bytes(s)
0x00007FFF8D673A1F (0x00007FF60BD2CCC8 0x00007FF60BD2CC40 0x0000002F00000119 0x00007FF600000000), _wassert() + 0x2F bytes(s)
0x00007FF60B973D4A (0x0000003768388F88 0x00000000FDFDFDFD 0x0000018D0000002F 0x0000003768388C01), llvm::SmallVectorTemplateCommon<clang::format::WhitespaceManager::Change,void>::operator[]() + 0x3A bytes(s), E:\src\llvm-project\llvm\include\llvm\ADT\SmallVector.h, line 281 + 0x2B byte(s)
0x00007FF60B972210 (0x0000003768388F88 0x0000003768388C08 0x0000003768388C28 0x00007FF600000005), clang::format::WhitespaceManager::getNetWidth<clang::format::WhitespaceManager::CellDescription *>() + 0x60 bytes(s), E:\src\llvm-project\clang\lib\Format\WhitespaceManager.h, line 267 + 0x13 byte(s)
0x00007FF60B9720FD (0x0000003768388F88 0x0000003768388D30 0x0000003768388C98 0xCCCCCC0000000005), clang::format::WhitespaceManager::getMaximumNetWidth<clang::format::WhitespaceManager::CellDescription *>() + 0x10D bytes(s), E:\src\llvm-project\clang\lib\Format\WhitespaceManager.h, line 304 + 0x1F byte(s)
0x00007FF60B96107F (0x0000003768388F88 0x0000003768388DD0 0x000000370000000C 0x00007FF600000036), clang::format::WhitespaceManager::alignArrayInitializersRightJustified() + 0x24F bytes(s), E:\src\llvm-project\clang\lib\Format\WhitespaceManager.cpp, line 1021 + 0x47 byte(s)
0x00007FF60B960DAF (0x0000003768388F88 0x000000000000000C 0x0000003700000036 0x0000003800000038), clang::format::WhitespaceManager::alignArrayInitializers() + 0x6F bytes(s), E:\src\llvm-project\clang\lib\Format\WhitespaceManager.cpp, line 988 + 0x32 byte(s)
0x00007FF60B960D00 (0x0000003768388F88 0x0000018D001D9730 0x0000003768389E10 0xCCCCCCCCCCCCCCCC), clang::format::WhitespaceManager::alignArrayInitializers() + 0xE0 bytes(s), E:\src\llvm-project\clang\lib\Format\WhitespaceManager.cpp, line 974
0x00007FF60B95ED7B (0x0000003768388F88 0x000000376838AFC8 0x0000003768388F00 0x0000003700000000), clang::format::WhitespaceManager::generateReplacements() + 0xAB bytes(s), E:\src\llvm-project\clang\lib\Format\WhitespaceManager.cpp, line 106
0x00007FF60B8E50F9 (0x000000376838B340 0x000000376838B0A8 0x000000376838B078 0x000000376838AFC8), clang::format::`anonymous namespace'::Formatter::analyze() + 0x399 bytes(s), E:\src\llvm-project\clang\lib\Format\Format.cpp, line 1860 + 0xA byte(s)
0x00007FF60B98A916 (0x000000376838B340 0x000000376838D3C8 0x000000376838C0D0 0x000000376838E0F8), clang::format::TokenAnalyzer::process() + 0x4F6 bytes(s), E:\src\llvm-project\clang\lib\Format\TokenAnalyzer.cpp, line 128
0x00007FF60B8ED165 (0x000000376838D260 0x000000376838D3C8 0x0000018D001B71C0 0xCCCCCCCCCCCCCCCC), <lambda_be157eee42e494a9ed92b4bd992e61c3>::operator()() + 0x65 bytes(s), E:\src\llvm-project\clang\lib\Format\Format.cpp, line 3060 + 0x3C byte(s)
0x00007FF60B90CE68 (0x000000376838D3C8 0x000000376838D260 0x0000018D001B71C0 0x00007FF60B928004), std::invoke<<lambda_be157eee42e494a9ed92b4bd992e61c3> &,clang::format::Environment const &>() + 0x28 bytes(s), D:\Program Files\Microsoft Visual Studio\2022\Preview\VC\Tools\MSVC\14.31.30818\include\type_traits, line 1490 + 0x14 byte(s)
0x00007FF60B8F9688 (0x000000376838D3C8 0x000000376838D260 0x0000018D001B71C0 0x0000000000000000), std::_Invoker_ret<std::pair<clang::tooling::Replacements,unsigned int>,0>::_Call<<lambda_be157eee42e494a9ed92b4bd992e61c3> &,clang::format::Environment const &>() + 0x28 bytes(s), D:\Program Files\Microsoft Visual Studio\2022\Preview\VC\Tools\MSVC\14.31.30818\include\functional, line 673 + 0x14 byte(s)
0x00007FF60B8EE217 (0x000000376838D258 0x000000376838D3C8 0x0000018D001B71C0 0xCCCCCCCC00000000), std::_Func_impl_no_alloc<<lambda_be157eee42e494a9ed92b4bd992e61c3>,std::pair<clang::tooling::Replacements,unsigned int>,clang::format::Environment const &>::_Do_call() + 0x37 bytes(s), D:\Program Files\Microsoft Visual Studio\2022\Preview\VC\Tools\MSVC\14.31.30818\include\functional, line 835 + 0x23 byte(s)
0x00007FF60B925BD5 (0x000000376838D258 0x000000376838D3C8 0x0000018D001B71C0 0x000000376838D750), std::_Func_class<std::pair<clang::tooling::Replacements,unsigned int>,clang::format::Environment const &>::operator()() + 0x65 bytes(s), D:\Program Files\Microsoft Visual Studio\2022\Preview\VC\Tools\MSVC\14.31.30818\include\functional, line 882 + 0x2C byte(s)
0x00007FF60B8E37A7 (0x000000376838D820 0x000000376838DA48 0x000000376838D840 0x000000376838D850), clang::format::internal::reformat() + 0xA27 bytes(s), E:\src\llvm-project\clang\lib\Format\Format.cpp, line 3078
0x00007FF60B8E0AE9 (0x000000376838E118 0x000000376838DA48 0x000000376838F540 0x000000376838F550), clang::format::reformat() + 0xB9 bytes(s), E:\src\llvm-project\clang\lib\Format\Format.cpp, line 3103 + 0x9C byte(s)
0x00007FF60B6A8E26 (0x000000376838FA50 0x0000018D001C4450 0x000000376838FA30 0x0000000000000000), clang::format::format() + 0x11C6 bytes(s), E:\src\llvm-project\clang\tools\clang-format\ClangFormat.cpp, line 481
0x00007FF60B6AA178 (0x0000000000000004 0x0000018D001CAF50 0x0000000000000000 0x00007FF60BB75C2D), main() + 0x5F8 bytes(s), E:\src\llvm-project\clang\tools\clang-format\ClangFormat.cpp, line 614 + 0x56 byte(s)
0x00007FF60BB76729 (0x00007FF60BD23000 0x00007FF60BD235A8 0x0000000000000000 0x0000000000000000), invoke_main() + 0x39 bytes(s), d:\a01\_work\14\s\src\vctools\crt\vcstartup\src\startup\exe_common.inl, line 79
0x00007FF60BB7660E (0x0000000000000000 0x0000000000000000 0x0000000000000000 0x0000000000000000), __scrt_common_main_seh() + 0x12E bytes(s), d:\a01\_work\14\s\src\vctools\crt\vcstartup\src\startup\exe_common.inl, line 288 + 0x5 byte(s)
0x00007FF60BB764CE (0x0000000000000000 0x0000000000000000 0x0000000000000000 0x0000000000000000), __scrt_common_main() + 0xE bytes(s), d:\a01\_work\14\s\src\vctools\crt\vcstartup\src\startup\exe_common.inl, line 331
0x00007FF60BB767BE (0x00000037678D2000 0x0000000000000000 0x0000000000000000 0x0000000000000000), mainCRTStartup() + 0xE bytes(s), d:\a01\_work\14\s\src\vctools\crt\vcstartup\src\startup\exe_main.cpp, line 17
0x00007FFFE2C07034 (0x0000000000000000 0x0000000000000000 0x0000000000000000 0x0000000000000000), BaseThreadInitThunk() + 0x14 bytes(s)
0x00007FFFE4442651 (0x0000000000000000 0x0000000000000000 0x0000000000000000 0x0000000000000000), RtlUserThreadStart() + 0x21 bytes(s)
```

**Comment by dvyukov** [comment]
Same crash here with the latest clang:

```
$ clang-format --version
clang-format version 15.0.0 (https://github.com/llvm/llvm-project.git 437d4e01fe4c057509dff30efd560049ad07bc99)

$ cat .clang-format 
BasedOnStyle: Google
DerivePointerAlignment: false
PointerAlignment: Left
IndentCaseLabels: false
AllowShortBlocksOnASingleLine: Empty
AllowShortFunctionsOnASingleLine: Empty
AllowShortIfStatementsOnASingleLine: false
AllowShortLoopsOnASingleLine: false
ConstructorInitializerAllOnOneLineOrOnePerLine: false
BreakConstructorInitializersBeforeComma: true
AlignArrayOfStructures: Left

$ clang-format --style=file test.cpp 
PLEASE submit a bug report to https://github.com/llvm/llvm-project/issues/ and include the crash backtrace.
Stack dump:
0.	Program arguments: clang-format --style=file test.cpp
 #0 0x00000000004519d3 llvm::sys::PrintStackTrace(llvm::raw_ostream&, int) (llvm/build/bin/clang-format+0x4519d3)
 #1 0x000000000044fdbe llvm::sys::RunSignalHandlers() (llvm/build/bin/clang-format+0x44fdbe)
 #2 0x0000000000451fba SignalHandler(int) Signals.cpp:0:0
 #3 0x00007f7583ed8200 __restore_rt (/lib/x86_64-linux-gnu/libpthread.so.0+0x13200)
 #4 0x00000000004c6620 unsigned int clang::format::WhitespaceManager::getMaximumNetWidth<clang::format::WhitespaceManager::CellDescription*>(clang::format::WhitespaceManager::CellDescription* const&, clang::format::WhitespaceManager::CellDescription* const&, unsigned int, unsigned int) const (llvm/build/bin/clang-format+0x4c6620)
 #5 0x00000000004c6054 clang::format::WhitespaceManager::alignArrayInitializersLeftJustified(clang::format::WhitespaceManager::CellDescriptions&&) (llvm/build/bin/clang-format+0x4c6054)
 #6 0x00000000004c4a63 clang::format::WhitespaceManager::alignArrayInitializers(unsigned int, unsigned int) (llvm/build/bin/clang-format+0x4c4a63)
 #7 0x00000000004c1d46 clang::format::WhitespaceManager::generateReplacements() (llvm/build/bin/clang-format+0x4c1d46)
 #8 0x0000000000479361 clang::format::(anonymous namespace)::Formatter::analyze(clang::format::TokenAnnotator&, llvm::SmallVectorImpl<clang::format::AnnotatedLine*>&, clang::format::FormatTokenLexer&) Format.cpp:0:0
 #9 0x0000000000497106 clang::format::TokenAnalyzer::process() (llvm/build/bin/clang-format+0x497106)
#10 0x0000000000486fad std::_Function_handler<std::pair<clang::tooling::Replacements, unsigned int> (clang::format::Environment const&), clang::format::internal::reformat(clang::format::FormatStyle const&, llvm::StringRef, llvm::ArrayRef<clang::tooling::Range>, unsigned int, unsigned int, unsigned int, llvm::StringRef, clang::format::FormattingAttemptStatus*)::$_10>::_M_invoke(std::_Any_data const&, clang::format::Environment const&) Format.cpp:0:0
#11 0x0000000000472412 clang::format::internal::reformat(clang::format::FormatStyle const&, llvm::StringRef, llvm::ArrayRef<clang::tooling::Range>, unsigned int, unsigned int, unsigned int, llvm::StringRef, clang::format::FormattingAttemptStatus*) (llvm/build/bin/clang-format+0x472412)
#12 0x0000000000473143 clang::format::reformat(clang::format::FormatStyle const&, llvm::StringRef, llvm::ArrayRef<clang::tooling::Range>, llvm::StringRef, clang::format::FormattingAttemptStatus*) (llvm/build/bin/clang-format+0x473143)
#13 0x000000000040c16b clang::format::format(llvm::StringRef) ClangFormat.cpp:0:0
#14 0x000000000040ab95 main (llvm/build/bin/clang-format+0x40ab95)
#15 0x00007f75839587ed __libc_start_main ./csu/../csu/libc-start.c:332:16
#16 0x0000000000407e0a _start (llvm/build/bin/clang-format+0x407e0a)
Segmentation fault

$ cat test.cpp 
TEST(Operation, TableTest) {
  struct Test {
    Op op;
    uptr a;
    uptr b;
    uptr c;
    uptr d;
    uptr e;
    uptr f;
    uptr g;
    uptr h;
  };
  Test tests[] = {
      {Or, 0, 0,       0},
      {Or,   0xab00,   0xcd, 0xabcd},
      {Or,   0x0f0f,   0x00ff, 0x0fff, 0xffff, 0xffff, 0xffff},
      {Or,   0x0f0f,   0x00ff, 0x0fff, 0x0000, 0xffff, 0xf0f0},
      {Or,   0x0f0f,   0x00ff, 0x0fff, 0xffff, 0x0000, 0xff00},
  };
}
```

**Comment by mydeveloperday** [comment]
This crash can be partially avoided by https://reviews.llvm.org/D121069 which will leave the structure alone when its non-square

**Comment by rymiel** [comment]
The crash was fixed by adfe58b09df957f8911c5e152daa9564e5931631, and while the format isn't ideal after that patch, this issue tracks a crash which has been fixed.
